### ISSUE TITLE
TaskQueue resolves before all tasks resolves

### ISSUE THREAD

**Issue by lojzatran** [sphereio/sphere-node-utils#32]
### Description
TaskQueue does not wait for all added tasks to finish and resolves too early. It's also not very clear from the documentation how I should add multiple tasks to the queue: https://github.com/sphereio/sphere-node-utils#taskqueue

### Expected Behavior
TaskQueue should wait for all tasks to finish before finish itself.

### Current Behavior
1. Add 5 tasks to TaskQueue: `[task1, task2, task3, task4, task5]`, where task3 takes longest time to finish.
2. When TaskQueue finishes, only 4 tasks are finished, task3 is still not finished. So the finished array is `[task1, task2, task4, task5]`
3. In order for everything to work correctly in the current implementation, user should have a reference to the longest running promise and only resolve after this promise is done. However, most of the time it's not possible to know which promise is the longest running.

### Possible Solution
TaskQueue should postpone resolving/rejecting Promises only after all of them are completed. With this approach, TaskQueue will return only after the longest running Promise in queue is finished.

### Test Case
```js
it 'should execute and finish all tasks', (done) ->
    expectedResult = [1, 2, 10, 3, 4]
    actualResult = []

    callMe = (input)->
      new Promise (resolve, reject) ->
        setTimeout ->
          actualResult.push(input)
          resolve true
        , 1000 * input

    taskPromise = null

    expectedResult.forEach((i) =>
      taskPromise = @task.addTask(callMe.bind(this, i))
    )

    taskPromise
      .then () ->
        expect(actualResult.length).toEqual(expectedResult.length)
        done()
      .catch (error) -> done(error)
```

**Comment by butenkor** [comment]
I think the idea of task queue is not to return when all tasks are finished but return finished tasks ASAP as queue might get pushed during execution new tasks in a loop. Did you try: 
```
promises = []
expectedResult.forEach((i) =>
      promises.push(@task.addTask(callMe.bind(this, i))
)

Promise.all(promises)
.then () ->
  expect(actualResult.length).toEqual(expectedResult.length)
  done()
.catch (error) -> done(error)
```

**Comment by lojzatran** [comment]
I tried, it does not work. In your case, we would have 5 promises in `promises` array? Actually my approach is similar to Promise.map() from bluebird. You need to wait until all promises are finished as well until you get the result. From Promise.map: `Promises returned by the mapper function are awaited for and the returned promise doesn't fulfill until all mapped promises have fulfilled as well.`

**Comment by lojzatran** [comment]
I found out the way how to use it with `Promise.all` according to sphere-node-sdk library. Therefore this bug is not valid.
